`arch/libcontext`: Re-factoring context, extended context, and TLS/TCB handling

skuenzer opened the pull request:
This PR is a preparation for refactoring the `uksched` API and for `clone` system call support (needed by `libmusl`). It revises the API for handling context, extended context (floating point, vector units, ...) and TLS/TCB. All of these interfaces are provided by a new architecture library called `libcontext`.

The following highlights are implemented by this PR: 
- Support for lightweight contexts (threads) with and without TLS/TCB
- Support for lightweight contexts (threads) that run only ISR-safe code
- Support for user-defined stack sizes
- Support for embedding custom TCBs by third-party libraries (e.g., libc, pthread-embedded)

Please note that this PR breaks `pthread-embedded`. It needs a reintegration later (after musl integration completed).

Special thanks to @kubanrob for the Aarch64 implementation of `libcontext`.

### Base target

 - Architecture(s): [`x86_64`] [`aarch64`]
 - Platform(s): any
 - Application(s): any

skuenzer committed:
plat/common: Remove unused `ukplat_ctx_type`

`ukplat_ctx_type` was intended to support CPU context saving and
restoring with different implementations at runtime (hw, sw). It
turned out that only the software context was effectively used and
that further abstraction is not necessary, even for preemptive
scheduling. This commit removes the context abstraction.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
plat/*: Handle TLS pointer separately from `ukplat_ctx` (API, x86)

Makes the platform context independent of thread-local storage (TLS).
This enables the construction of schedulers supporting threads with and
without TLS.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uksched: Save current scheduled thread in a variable

Saves the current scheduled thread reference in a variable as part
of `lib/uksched`. This actually solves two limitations:
1) By not storing the current thread reference on the top of the current
   thread stack, custom sized stacks can be supported. There is also no
   platform-specific implementation required anymore.
2) By not relying on the TLS/TCB pointer, the Unikraft architecture
   remains independent from later TCB implementation where the layout
   might be defined by a libc or pthread library. In addition,
   lightweight threads without TLS/TCB can be created and scheduled, as
   well.

With this change, the `ukplat_stack_set_current_thread()` helper is
removed as well. As soon as CPU-local variable declaration is available,
the variable should be defined per-CPU.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
arch/libcontext: API, x86: ctx handling and switching

This commit moves CPU context handling (context initialization and
context switch logic) from the platform libraries to the architecture
library `libcontext`.
Additionally, the following API refactoring and additions are provided:
- Helpers to initialize a new context (`ukarch_ctx`) with a context
  function or instruction pointer (IP)
- The possibility to clear the standard register set to `0x0` before
  entering the context function or given IP.
- There is no more context encapsulation. This means that context
  functions should not return. Context exit functions can later be
  provided via a scheduler implementation.
  This guarantees that context execution can even be started at any code
  location with an unchanged stack pointer, which is necessary to
  provide functionality like the `clone` system call.

The code is compiled isr-safe so that the compiler will never use
extended CPU units for the context initialization and switch operation.
This allows constructing lightweight context switches to code that does
not use extended registers.
`lib/uksched` is modified accordingly to use this new API.

Checkpatch-Ignore: LONG_LINE_STRING
Checkpatch-Ignore: AVOID_EXTERNS
Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
arch/libcontext: API, x86: ectx saving and loading

This commit introduces a new architecture library: `libcontext`. Since
saving and restoring of execution context is actually platform-
independent, this and the subsequent commits will migrate the
corresponding functionality from the platforms to `libcontext`.

This commit migrates the loading and saving of extended CPU unit context
(such as vector, floating point) to `libcontext`. Extended context
saving is handled independently from `ukplat_ctx`.

Checkpatch-Ignore: INITIALISED_STATIC
Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
arch/libcontext: API, x86: Helper for retrieving a stack pointer

Because stack growth direction and alignment are defined by the CPU
architecture, this commits provides a helper to derive a valid stack
pointer for a given stack memory allocation.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
arch/libcontext: API, x86: TLS initialization

This commit moves the TLS initialization code into the architecture's
library `libcontext`. This commit also fixes the TLS allocation problem
caused by incorrect padding which was introduced by PR #287,
commit 5988db03a3af ("arch/x86_64: increase TLS alignment").

Checkpatch-Ignore: LINE_CONTINUATIONS
Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

kubanrob committed:
arch/libcontext: aarch64: Implementation for (e)ctx

Changes made in this commit:

arch/arm/arm64/ctx.S:
- import `asm_ctx_switch` from `plat/common/arm/thread_start64.S`
  as `ukarch_ctx_switch`
- translate x86_64 version for `_ctx_arm_callN` and `_ctx_arm_clearregs`

arch/arm/arm64/include/uk/asm/ctx.h, arch/arm/ctx.c:
- import x86_64 version
- modify header for 16 byte stack alignment

arch/arm/ectx.c:
- import from x86_64 version
- modify to use implementation from `arm/arm64/cpu.h`

plat/common/include/arm/arm64/cpu.h:
- adapt for use with ectx

Checkpatch-Ignore: AVOID_EXTERNS
Signed-off-by: Robert Kuban <robert.kuban@opensynergy.com>

kubanrob committed:
arch/libcontext: aarch64: Implementation for TLS

Changes made in this commit:

arch/arm/arm64/tls.c:
- import from x86_64 implementation
- adapt TLS layout to aarch64

plat/common/include/arm/arm64/tls.h:
- implement by storing to / loading from the `tpidr_el0` register

Checkpatch-Ignore: AVOID_EXTERNS
Signed-off-by: Robert Kuban <robert.kuban@opensynergy.com>

skuenzer committed:
arch/libcontext: API, x86: Custom TCB as part of a TLS allocation

Allows reserving space for a user-defined TCB in the TLS memory area.
This functionality is enabled as soon as one (but only one) library
reserves space for a custom TCB with the `Makefile.uk` function
`ukarch_tls_tcb_reserve()`. Such a library (e.g., pthread-embedded,
musl) must provide the `ukarch_tls_tcb_init()` function symbol, which is
called by `libcontext` whenever a TLS region is initialized.

Checkpatch-Ignore: CONSTANT_COMPARISON
Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

kubanrob committed:
arch/aarch64: Disable outline atomic

Quick-fix based on https://github.com/unikraft/unikraft/issues/319

aarch64-none-elf has the TLS disabled in the gcc configuration,
so aarch64-none-linux-gnu needs to be used for TLS support.

In case it is decided to implement a static version of emutls,
the patch can be reverted to enable the use of the bare-metal toolchain.

Signed-off-by: Robert Kuban <robert.kuban@opensynergy.com>

Commit:
arch/libcontext: aarch64: Use SP pointer instead of X2 when storing ctx

Let's consider context switch routine

ENTRY(asm_sw_ctx_switch)
    /* Save callee-saved registers to prevctx's stack */
    sub x2, sp, #__CALLEE_SAVED_SIZE
    stp x19, x20, [x2, #16 * 0]
    ...
    stp x29, x30, [x2, #16 * 5]
    (point 1) <-----------------------------------
    /*
     * Record the restore point for switch out thread to restore
     * its called-saved registers in next switch to time.
     */
    ldr x30, =restore_point
    /* Save sp and restore point to previous context */
    stp x2, x30, [x0]

When we save context, sp is not allocated (not decremented), because we use
intermediate x2 register. So at point1 the stack picture would be
  (addrs grow down)

   x19   <----- sp
   x20
    ..
   x30   <------x2

however, if the irq will occur at point 1, it will corrupt saved context
with the trapframe data, because it uses the same stack and sp was not
allocated.

Signed-off-by: Aleksandr Iashchenko <aia@opensynergy.com>

skuenzer committed:
include/plat: TLS execution helper wrapper

Provides a wrapper that allows a particular function to be executed with
a different TLS pointer enabled. This wrapper solves the problem that
the GCC compiler is not aware when the TLS register is changed. Because
of this, GCC would make incorrect assumptions about TLS variables and
might try to optimize loading and saving a TLS variable at an earlier
or later time within a function, ignoring that the TLS base register
might have changed.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

kubanrob committed:
arch/libcontext: aarch64: Handle x18 register during context switch

The role of register x18 is platform specific, but saving and restoring
it later is safer then leaving random values in it.

Signed-off-by: Robert Kuban <robert.kuban@opensynergy.com>

kubanrob committed:
arch/libcontext: aarch64: Compiler barrier for `set_tls_pointer()`

This commit adds a compiler barrier to `set_tls_pointer()`, which sets
the TLS register. This avoids re-ordering.

Checkpatch-Ignore: COMPLEX_MACRO
Signed-off-by: Robert Kuban <robert.kuban@opensynergy.com>

dragosargint reviewed (commented):
Hi, Simon

I went through this PR, things seem in order. But the problem is that it breaks `pthread_embedded`. So either a fix for `pthread-embedded` must be taken into account, or the base of the PR must be changed, or we just accept the fact that `pthread-embedded` is broken.

dragosargint commented on lib/uksched/thread.c:
```diff
@@ -41,27 +41,21 @@
 
-static void init_sp(unsigned long *sp, char *stack,
-		void (*function)(void *), void *data)
+static __noreturn void thread_capsule(long funcp, long argp)
```
Now here is what I found, this is how a thread is created for `pthread-embedded`
```
struct pte_entry_capsule {
        pte_osThreadEntryPoint entry_point;
        void *argv;
};
...
pte_osThreadCreate() { 
  ...
  capsule.entry_point = entry_point;
  capsule.argv        = argv;
  ...
  th = uk_thread_create_attr(NULL, NULL, PTE_CAPSULE_MAGIC, &capsule);
  ...
}
```

When the execution reaches this function (i.e. `thread_capsule()`), the function pointer will point to `PTE_CAPSULE_MAGIC`, which is just garbage in this case.

skuenzer replied:
You are right, in fact, we currently decided to break `newlib` and `pthread-embedded`. The idea is to come back to it later, this is why we have currently `staging-muslpreview` as branch for these PRs. `newlib` and especially `pthread-embedded` need an updated integration. However, this would be much cleaner with the latest APIs.
Additionally, this capsule is a temporary solution for implementing libcontext with the current `uksched` API. The upcoming scheduler PR with the new API will properly integrate `libcontext` and will make this capsule unnecessary.

skuenzer committed:
plat/linuxu: Include TLS, eventtab, and exception sections for Arm

Includes our format of the static TLS section, eventtab and exception
sections to linuxu builds for Arm (arm32, and aarch64).

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
plat/linuxu: Enable scheduling for aarch64

In order to enable scheduling support, the common aarch64 TLS pointer
set and retrieval routines are used. Using the `arch_prctl` system call
to set and get the TLS pointer register value is specific to Linux
userspace on x86.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

dragosargint reviewed (approved):
This looks great!
Reviewed-by: Dragos Iulian Argint <dragosargint21@gmail.com>

razvand commented on arch/arm/arm64/ctx.S:
```diff
@@ -0,0 +1,161 @@
+	 *
+	 * The role of register x18 is platform specific,
+	 * but saving and restoring it later is definitely
+	 * safer then leaving random values in it.
```
```suggestion
	 * safer than leaving random values in it.
```

skuenzer replied:
True, fixed.

razvand commented on arch/arm/arm64/ctx.S:
```diff
@@ -0,0 +1,161 @@
+         */
+	str x18, [sp, #-16]!
+
+	/* Save callee-saved registers to prevctx's stack */
```
```suggestion
	/* Save callee-saved registers to prevctx's stack. */
```

skuenzer replied:
I think the dot at the end of one-liner comment is really only personal preference. Since the file has more one-liner comments without ending dot, I'd leave it as is.

razvand commented on arch/arm/arm64/ctx.S:
```diff
@@ -0,0 +1,161 @@
+	ret
+
+/*
+ * Do a local context switch taken from thread_start64.S
```
I don't understand the wording _taken from thread_start64.S_

kubanrob replied:
This code was adapted from the code in `plat/common/arm/thread_start64.S`. The author / copyright information in the header also reflect that.

Probably there is a better way to annotate that.

skuenzer replied:
It is confusing, I updated the comment to
```
/*
 * Switch context on the current LCPU.
 */
```

razvand commented on arch/arm/arm64/include/uk/asm/ctx.h:
```diff
@@ -0,0 +1,65 @@
+#error Do not include this header directly
+#endif
+
+/* stack needs to be quadword (16 byte) aligned */
```
```suggestion
/* Stack needs to be quadword (16 bytes) aligned. */
```

razvand replied:
Isn't quadword 8 bytes?

kubanrob replied:
Wordsize is depended on the architecture on ARM, its 32bits (see: https://stackoverflow.com/a/55430777). However, I'm not opposed to drop the "quadword" completely in favor of "16 bytes".

skuenzer replied:
I am removing the word `quadword` to avoid confusion.

razvand commented on arch/arm/arm64/tls.c:
```diff
@@ -0,0 +1,281 @@
+#endif /* !CONFIG_LIBUKDEBUG */
+
+/*
+ * aarch64 ABI requires 16 bytes of TCB
+ * after the TLS arch pointer
```
```suggestion
 * aarch64 ABI requires 16 bytes of TCB after the TLS arch pointer.
```

skuenzer replied:
Fixed

razvand commented on arch/arm/arm64/tls.c:
```diff
@@ -0,0 +1,281 @@
+ * aarch64 ABI requires 16 bytes of TCB
+ * after the TLS arch pointer
+ *
+ * - in musl that included in the size of the TCB (ptread)
+ * - other libc might need to add 16 byte after the TCB
```
I'm not sure I understand these two sentences. @kubanrob , could you assist?

kubanrob replied:
It is explained here very well: https://www.openwall.com/lists/musl/2018/02/09/7

Basically: aarch64 ABI demands a block of 16 bytes after the tlsp, which demands values to be stored there. However, do to the structure of the TLS these values contain not enough to access the TLS and so the compiler generated code never accesses this block. Knowing that, musl just reuses that space for other members in the TCB.

Basically these 16bytes are part of the TCB in this layout. If you don't want them to be for some libc, you need to add 16bytes to the actual size of the TCB.

skuenzer replied:
I changed the whole block to:
```c
/*
 * Aarch64 ABI requires 16 bytes of space after the tlsp as part of the
 * TCB.
 * NOTE: We assume that libraries (like musl) reserve this space when using
 *       this TLS API. You can find more information here:
 *       https://www.openwall.com/lists/musl/2018/02/09/7
 */
```

razvand commented on arch/arm/arm64/tls.c:
```diff
@@ -0,0 +1,281 @@
+/*
+ * This file implements a static exec TLS layout
+ * (variant 2 with one static TLS block).
+ * The calls create a TLS containing space for a TCB.
```
What are "The calls"? Why plural?

kubanrob replied:
This introduction was copied from `arch/x86/x86_64/tls.c`. I would not mind dropping the sentence.

skuenzer replied:
Removed

razvand commented on arch/arm/arm64/tls.c:
```diff
@@ -0,0 +1,281 @@
+ *             +-------------------------+ /  /
+ *
+ * Note: ukarch_tlsp_get() points to the last 16 bytes of TCB of the thread.
+ *       AARCH64 ABI requieres the tlsp to point to 2 pointers (16 bytes),
+ *       the pointer to the DTV and the an implementation specific pointer.
```
```suggestion
 *       AARCH64 ABI requires the tlsp to point to 2 pointers (16 bytes):
 *       the pointer to the DTV and the an implementation specific pointer.
```

skuenzer replied:
Fixed

razvand commented on arch/arm/arm64/tls.c:
```diff
@@ -0,0 +1,281 @@
+ *       AARCH64 ABI requieres the tlsp to point to 2 pointers (16 bytes),
+ *       the pointer to the DTV and the an implementation specific pointer.
+ *
+ *       In reality, this 16 byte area is never used by the compiler
```
```suggestion
 *       In reality, this 16 bytes area is never used by the compiler
```

skuenzer replied:
Fixed

razvand commented on arch/arm/arm64/tls.c:
```diff
@@ -0,0 +1,281 @@
+ */
+
+/*
+ * this is usually read from the elf header
+ * (p_align of the TLS section)
+ *
+ * this NEEDS to match, as aarch64 uses
+ * this at link time to calculate the offset
+ * from the tlsp
```
```suggestion
 * This is usually read from the ELF header (p_align of the TLS section).
 *
 * This NEEDS to match, as aarch64 uses this at link time to calculate the offset
 * from the tlsp.
```

skuenzer replied:
Fixed

razvand commented on arch/arm/arm64/tls.c:
```diff
@@ -0,0 +1,281 @@
+static __sz tls_data_offset(void)
+{
+	/*
+	 * the distance between tlsp and the start of
+	 * the initial TLS segment is max(16, p_align)
+	 * (source:
+	 * https://fuchsia.dev/fuchsia-src/development/kernel/threads/tls)
```
```suggestion
	 * The distance between tlsp and the start of the initial TLS segment is max(16, p_align).
	 * (source: https://fuchsia.dev/fuchsia-src/development/kernel/threads/tls)
```

skuenzer replied:
Fixed

razvand commented on arch/arm/arm64/tls.c:
```diff
@@ -0,0 +1,281 @@
+static __sz tls_padding_2_size(void)
+{
+	/*
+	 * The tlsp points to the mandatore 16 byte block,
+	 * padding 2 fills the rest of it.
```
```suggestion
	 * The tlsp points to the mandatory 16 bytes block, padding 2 fills the rest of it.
```

skuenzer replied:
Updated

razvand commented on arch/arm/arm64/tls.c:
```diff
@@ -0,0 +1,281 @@
+	 * The tlsp points to the mandatore 16 byte block,
+	 * padding 2 fills the rest of it.
+	 *
+	 * The size of this padding is therefore determined
+	 * by the ABI.
```
```suggestion
	 * The size of this padding is therefore determined by the ABI.
```

skuenzer replied:
Updated

razvand commented on arch/arm/arm64/tls.c:
```diff
@@ -0,0 +1,281 @@
+	 * - TCB_SIZE
+	 * - TLS_AREA_ALIGN
+	 * - and therefore size of padding 2
+	 * but adding data_offset don't break alignment,
+	 * so we can just align tlsp.
```
```suggestion
	 * but adding data_offset doesn't break alignment, so we can just align tlsp.
```
What is `data_offset`?

kubanrob replied:
It is `tls_data_offset`, which is also in the shown in the diagram above.

skuenzer replied:
Updated

razvand commented on arch/arm/ctx.c:
```diff
@@ -0,0 +1,143 @@
+
+	UK_ASSERT(ctx);
+	UK_ASSERT(sp);			/* a stack is needed */
+	UK_ASSERT(ip);			/* NULL as IP will cause a crash */
```
```suggestion
	UK_ASSERT(ip);			/* IP == NULL will cause a crash */
```

kubanrob replied:
This file is a copy of `arch/x86/ctx.c`, I forgot to mentioned that up to now ... maybe we can put this file in a common directory and fix problems in one place @skuenzer.

skuenzer replied:
I added a comment regarding the origin of this file

razvand commented on arch/arm/ctx.c:
```diff
@@ -0,0 +1,143 @@
+	UK_ASSERT(ctx);
+	UK_ASSERT(sp);			/* a stack is needed */
+	UK_ASSERT(ip);			/* NULL as IP will cause a crash */
+	UK_ASSERT(!keep_regs && sp);	/* a stack is needed when clearing */
```
Clearing what?

skuenzer replied:
Standard registers... I added this to the comment.

razvand commented on arch/arm/ctx.c:
```diff
@@ -0,0 +1,143 @@
+
+	UK_ASSERT(ctx);
+	UK_ASSERT(sp);			/* a stack is needed */
+	UK_ASSERT(entry);		/* NULL as func will cause a crash */
```
```suggestion
	UK_ASSERT(entry);		/* entry == NULL will cause a crash */
```

skuenzer replied:
Hum... 🤔 I assume this is more personal preference. I am actually fine with both ways (original and your suggestion). I updated it to reflect that we speak about the entry parameter here.

razvand commented on include/uk/arch/ctx.h:
```diff
@@ -0,0 +1,234 @@
+} __packed;
+
+/*
+ * Context functions are not allowed to return
```
```suggestion
 * Context functions are not allowed to return.
```

skuenzer replied:
Updated

razvand reviewed (commented):
Hi, @skuenzer . I did a first pass and made some cosmetic comments.

I'm not that fond of very limited comments. I thin they are often limited to 50 columns, which is pretty darn ugly and hard to follow. I adhere to a more lax coding style with respect to column limits, such as [Linus Torvalds]() pointed out.

Also, I'm assuming the internal functions don't need Doxygen-style comments right?

Do you want me to be pedantic about Doxygen-style comments or we can come to that at some point in the future? (to hasten this PR being merged)

I will do a more in-depth pass + testing in the next few days.

gabrielmocanu commented:
@skuenzer 
Can you please provide an update on this PR? Thanks!

skuenzer commented:
> @skuenzer Can you please provide an update on this PR? Thanks!

I just pushed the update. Thanks!

adinasm commented on include/uk/arch/ctx.h:
```diff
@@ -0,0 +1,234 @@
+			    ukarch_ctx_entry2 entry, long arg0, long arg1);
+
+/**
+ * Pushes a value to the stack of an remote context that should
```
```suggestion
 * Pushes a value to the stack of a remote context that should
```

skuenzer replied:
Thanks a lot for spotting this 👍🏻, I fixed it.

razvand commented on lib/uksched/sched.c:
```diff
@@ -47,6 +45,9 @@
 
 struct uk_sched *uk_sched_head;
 
+/* TODO: Define per CPU */
```
When should this TODO item be fixed, i.e define `__uk_sched_thread_current` as a per-CPU variable?

skuenzer replied:
This is needs to done as soon as we have a per-CPU variable declaration. It will be required to support scheduling on multiple cores.

skuenzer replied:
I added some more comments to the commit message and to the comment in the source.

razvand commented on arch/Makefile.rules:
```diff
@@ -0,0 +1,14 @@
+_UKARCH_TLS_HAVE_TCB=
+
+## Make sure that the reserved space for a TCB is set only once.
+## A second call of this function will make compiling stop.
+## Usage:
+##  $(eval $(call ukarch_tls_tcb_reserve,<tcb_size_in_bytes>))
```
Any reason to use double comment signs (`##`) instead of a single one? Is this something to insist on?

skuenzer replied:
Good question. It seems that I took it over from uknetdev but it is obviously not needed. I'll make it to single `#`.

razvand commented on plat/common/arm/thread_start64.S:
```diff
@@ -29,7 +29,7 @@
  * POSSIBILITY OF SUCH DAMAGE.
  */
-#include <uk/plat/common/sw_ctx.h>
+//#include <uk/plat/common/ctx.h>
```
Shouldn't this simply be removed?

skuenzer replied:
Yes, it should 😉

razvand commented on plat/common/include/arm/arm64/tls.h:
```diff
@@ -30,6 +31,13 @@
 #define __PLAT_CMN_ARM64_TLS_H__
 
-#define set_tls_pointer(ptr) do { } while (0)
+#include <arm/arm64/cpu.h> /* rdmsrl, wrmsrl */
```
```suggestion
#include <arm/arm64/cpu.h> /* SYSREG_READ, SYSREG_WRITE */
```

skuenzer replied:
:+1:

razvand commented on plat/linuxu/Config.uk:
```diff
@@ -1,7 +1,7 @@
        bool "Linux user space"
        default n
-       depends on !HAVE_SYSCALL && !HAVE_SMP && ((!ARCH_ARM_32 && !ARCH_ARM_64) || !HAVE_SCHED)
+       depends on !HAVE_SYSCALL && !HAVE_SMP && (!ARCH_ARM_32 || !HAVE_SCHED)
```
Why is `ARCH_ARM64` removed?

skuenzer replied:
Because with the changes, the linux user space target finally works also for Aarch64. It is making `linuxu` available when Arm is selected under architecture configuration.

razvand reviewed (changes requested):
Hi, @skuenzer , see my comments.

Any idea why there are [failing checks](https://github.com/unikraft/unikraft/pull/535/checks) from Codacy? I couldn't figure it out.

Also, please fix the wording of the message for [commit `a06dc8`](https://github.com/unikraft/unikraft/pull/535/commits/a06dc8e91d0c78ff89b51d2050af070f6db9c387). I would use:

```
This change avoids reordering when setting the TLS register.
```

skuenzer commented:
@razvand I just updated the PR. Thanks a lot for your review.

kubanrob commented on include/uk/plat/tls.h:
```diff
@@ -73,4 +80,43 @@ static inline void ukplat_tls_set(void *tls_area)
+ * This wrapper function will apply a given TLS pointer before executing
+ * a given function. After the execution, the original TLS pointer is
+ * restored.
+ * NOTE: This function is decalred as `__noinline`  so that we keep
```
`declared`

kubanrob commented on include/uk/arch/ctx.h:
```diff
@@ -35,10 +35,161 @@
+	})
+
+/**
+ * Switch the currrent logical CPU to context `load`. The current context
```
`current`

kubanrob commented on include/uk/arch/ctx.h:
```diff
@@ -35,10 +35,161 @@
+/**
+ * Pushes a value to the stack of a remote context that should
+ * not be executed at the moment. This macro reserves exactly
+ * the size of `value`'s datatype on the stack.
```
This comment is not true for aarch64, where the stack is aligned to 16 bytes in `ukarch_rstack_push` (7ecc8534ce30823e0efceca5bb0be8df0cef94c9).
This stack alignment is mandated by the architecture.

kubanrob reviewed (commented):
I found some spelling errors (mostly my own).

- commit message of 0b2a31da7953682bfa038ff680e96ad9cd2c06d5: adapt TLS layout to ~to~ aarch64
- commit message of 89349ff6f5c831f62c36e93a3c0074a948b7343c: `adopt` -> `adapt`, `adoption` -> `adaption`
I think it would be better to rephrase it completely:  ```This commit adds a compiler barrier to `set_tls_pointer()`, which sets the TLS register. This avoids re-ordering.```

skuenzer commented:
Hi @kubanrob , thanks a lot for going over it as well. I implemented the changes. :+1:

kubanrob reviewed (approved):
Thank you @skuenzer!

Reviewed-by: Robert Kuban <robert.kuban@opensynergy.com>

razvand reviewed (approved):
Thanks, @skuenzer 

Reviewed-by: Razvan Deaconescu <razvand@unikraft.io>
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
